src/translator/service.cpp

```diff
@@ -61,17 +61,26 @@ AsyncService::AsyncService(const AsyncService::Config &config)
   }
 }
 
-AsyncService::~AsyncService() {
+void AsyncService::terminate() {
+  safeBatchingPool_.clear();
+  join();
+}
+
+void AsyncService::join() {
   safeBatchingPool_.shutdown();
   for (std::thread &worker : workers_) {
     assert(worker.joinable());
```

Comment:
We compile with DNDEBUG, so this doesn't do anything in our release build. Do we assume threads are always joinable (Which I think is reasonable)

Reply:
Should we make some mechanism to ensure that this isn't called twice so we don't try to join already joined threads?

Reply:
Last line of `join()` is `workers.clear()` which will make sure that a second call to `join()` won't wait for already joined workers.

There could only be non-joinable threads in there if `join()` didn't clear `workers`, or `workers` somehow contained default-constructed threads. Both would we a programming error, not something that should be able to occur in runtime. I think using `assert()` is fine here (and otherwise it will just be `thread.join()` throwing an `invalid_argument` exception.)

Reply:
join doesn't hold a mutex so one could technically call it from two different threads before `workers_.clear()` is called. Do we assume this isn't ever going to happen?

Reply:
Hmm, that's right. I would expect any sane program to only ever call join or terminate from the same thread that created (and owns) the AsyncService instance. In that way it's similar to `std::thread::join`: *No synchronization is performed on \*this itself. Concurrently calling join() on the same thread object from multiple threads constitutes a data race that results in undefined behavior.*